rust/projects/parallelism/src/thread_pool/shared_queue.rs

```diff
@@ -0,0 +1,64 @@
+use std::{panic, thread};
+
+use super::ThreadPool;
+use crate::Result;
+
+use crossbeam::channel::{self, Receiver, Sender};
+
+// Note for Rust training course: the thread pool is not implemented using
+// `catch_unwind` because it would require the task to be `UnwindSafe`.
+
+/// A thread pool using a shared queue inside.
+///
+/// If a spawned task panics, the old thread will be destroyed and a new one will be
+/// created. It fails silently when any failure to create the thread at the OS level
+/// is captured after the thread pool is created. So, the thread number in the pool
+/// can decrease to zero, then spawning a task to the thread pool will panic.
+pub struct SharedQueueThreadPool {
+    tx: Sender<Box<dyn FnOnce() + Send + 'static>>,
+}
+
+impl ThreadPool for SharedQueueThreadPool {
+    fn new(threads: u32) -> Result<Self> {
+        let (tx, rx) = channel::unbounded::<Box<dyn FnOnce() + Send + 'static>>();
+        for _ in 0..threads {
+            let rx = rx.clone();
+            thread::Builder::new().spawn(move || run_tasks(rx))?;
+        }
+        Ok(SharedQueueThreadPool { tx })
+    }
+
+    /// Spawns a function into the thread pool.
+    ///
+    /// # Panics
+    ///
+    /// Panics if the thread pool has no thread.
+    fn spawn<F>(&self, job: F)
+    where
+        F: FnOnce() + Send + 'static,
+    {
+        self.tx
+            .send(Box::new(job))
+            .expect("The thread pool has no thread.");
+    }
+}
+
+fn run_tasks(rx: Receiver<Box<dyn FnOnce() + Send + 'static>>) {
+    let rx2 = rx.clone();
+    panic::set_hook(Box::new(move |panic_info| {
+        error!("Thread panicked: {}", panic_info);
+        let rx = rx2.clone();
+        if let Err(e) = thread::Builder::new().spawn(move || run_tasks(rx)) {
+            error!("Failed to spawn a thread: {}", e);
+        }
+    }));
```

Comment:
The panic hook shouldn't be used this way. There is only a single panic hook globally, so what's happening here is that the same hook is being recreated for each thread; and if any thread in the entire program panick's the thread pool is growing.

It's pretty much always the application's responsibility to set the panic hook, not any individual component's.

There are many other ways to replenish the pool on panic without using catch_unwind.

One would be to keep an atomic counter of threads that is increased on spawn and decreased on a custom drop, and make the ThreadPool object responsible for replenishing the pool.

Another would be to again use a custom drop in the thread along with `thread::panicking` to have the thread replenish itself like you are doing here.